File: src/main/linkedlist/java/ExecTime.java
import java.io.*;
import java.util.*;

public class ExecTime {
    public static void main(String[] args) {
        try {
            String[] methods = {
                "add_first", "add_middle", "add_last",
                "get_first", "get_middle", "get_last",
                "remove_first", "remove_middle", "remove_last"
            };

            BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
            String line;

            while ((line = reader.readLine()) != null) {
                if (line.isBlank()) continue;

                int[] input = Arrays.stream(line.split(" ")).mapToInt(Integer::parseInt).toArray();

                LinkedList ll = new LinkedList();
                for (int value : input) {
                    ll.addLast(value);
                }

                for (String method : methods) {
                    long[] results = new long[101];
                    for (int i = 0; i < 101; i++) {
                        results[i] = medeTempo(ll, method, input);
                    }
                    Arrays.sort(results);
                    long median = results[50];
                    salvaResultado(method, median, input.length);
                }
            }
        } catch (IOException ioe) {
            ioe.printStackTrace();
        }
    }

    private static long medeTempo(LinkedList ll, String method, int[] input) {
        int index = input.length / 2;
        long start = 0;
        long end = 0;

        switch (method) {
            case "add_first":
                start = System.nanoTime();
                ll.addFirst(10);
                end = System.nanoTime();
                ll.removeFirst();
                break;

            case "add_middle":
                start = System.nanoTime();
                ll.add(10, index);
                end = System.nanoTime();
                ll.removeByIndex(index);
                break;

            case "add_last":
                start = System.nanoTime();
                ll.addLast(10);
                end = System.nanoTime();
                ll.removeLast();
                break;

            case "get_first":
                start = System.nanoTime();
                ll.getFirst();
                end = System.nanoTime();
                break;

            case "get_middle":
                start = System.nanoTime();
                ll.get(index);
                end = System.nanoTime();
                break;

            case "get_last":
                start = System.nanoTime();
                ll.getLast();
                end = System.nanoTime();
                break;

            case "remove_first":
                int first = ll.getFirst();
                start = System.nanoTime();
                ll.removeFirst();
                end = System.nanoTime();
                ll.addFirst(first);
                break;

            case "remove_middle":
                int valueAtIndex = ll.get(index);
                start = System.nanoTime();
                ll.removeByIndex(index);
                end = System.nanoTime();
                ll.add(valueAtIndex, index);
                break;

            case "removeLast":
                int last = ll.getLast();
                start = System.nanoTime();
                ll.removeLast();
                end = System.nanoTime();
                ll.addLast(last);
                break;
        }

        return end - start;
    }

    private static void salvaResultado(String method, long tempo, int tamanhoEntrada) {
        String outputLine = "linkedlist_java " + tempo + " " + tamanhoEntrada;
        File file = new File(method + ".data");
        boolean isNewFile = !file.exists() || file.length() == 0;

        try (PrintWriter writer = new PrintWriter(new FileOutputStream(file, true))) {
            if (isNewFile) {
                writer.println("estrutura_linguagem tempo tamanho");
            }
            writer.println(outputLine);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/linkedlist/python/ExecTime.py
import sys
import time
import os
import statistics
from linked_list import LinkedList

methods = [
    "add_first", "add_middle", "add_last",
    "get_first", "get_middle", "get_last",
    "remove_first", "remove_middle", "remove_last"
]

def mede_tempo(ll, method, n):
    index = n // 2
    start = end = 0

    if method == "add_first":
        start = time.perf_counter_ns()
        ll.add_first(10)
        end = time.perf_counter_ns()
        ll.remove_first()

    elif method == "add_middle":
        start = time.perf_counter_ns()
        ll.add(10, index)
        end = time.perf_counter_ns()
        ll.remove_by_index(index)

    elif method == "add_last":
        start = time.perf_counter_ns()
        ll.add_last(10)
        end = time.perf_counter_ns()
        ll.remove_last()

    elif method == "get_first":
        start = time.perf_counter_ns()
        ll.get_first()
        end = time.perf_counter_ns()

    elif method == "get_middle":
        start = time.perf_counter_ns()
        ll.get(index)
        end = time.perf_counter_ns()

    elif method == "get_last":
        start = time.perf_counter_ns()
        ll.get_last()
        end = time.perf_counter_ns()

    elif method == "remove_first":
        val = ll.get_first()
        start = time.perf_counter_ns()
        ll.remove_first()
        end = time.perf_counter_ns()
        ll.add_first(val)

    elif method == "remove_middle":
        val = ll.get(index)
        start = time.perf_counter_ns()
        ll.remove_by_index(index)
        end = time.perf_counter_ns()
        ll.add(val, index)

    elif method == "remove_last":
        val = ll.get_last()
        start = time.perf_counter_ns()
        ll.remove_last()
        end = time.perf_counter_ns()
        ll.add_last(val)

    return end - start


def salva_resultado(method, tempo, tamanho):
    file_name = f"{method}.data"
    is_new_file = not os.path.exists(file_name) or os.stat(file_name).st_size == 0

    with open(file_name, "a") as file:
        if is_new_file:
            file.write("estrutura_linguagem tempo tamanho\n")
        file.write(f"linkedlist_python {tempo} {tamanho}\n")


def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        input_list = list(map(int, line.split()))
        n = len(input_list)

        ll = LinkedList()
        for num in input_list:
            ll.add_last(num)

        for method in methods:
            resultados = []
            for _ in range(101):
                tempo = mede_tempo(ll, method, n)
                resultados.append(tempo)

            mediana = statistics.median(resultados)
            salva_resultado(method, int(mediana), n)


if __name__ == "__main__":
    main()
